### PR TITLE
Added a role called tenant_admin to our deployment

### PR DESCRIPTION
**Pull request by ajayaa**

### CODE REVIEW

**Comment by bodepd** [comment]
this is definitely the wrong place for this code. it should go in `rjil::openstack_objects`

**Comment by ajayaa** [comment]
please test this

**Comment by bodepd** [comment]
@ajayaa can you rebase this to a single commit please. Also a little context in the commit message to why this user is required and why the existing tenant is not sufficient would be great.

thanks!

**Comment by ajayaa** [comment]
please test this

**Comment by ajayaa** [comment]
please test this

File: manifests/keystone/role.pp [hunk 1]
```diff
@@ -0,0 +1,15 @@
+##
+# Define: rjil::keystone::role
+#
+# == Purpose: create Keystone roles
+#
+##
+define rjil::keystone::role (
+  $role,
```
Comment (anshprat) [inline manifests/keystone/role.pp]: provide a default value (undef)

**Comment by anshprat** [comment]
Also, would be great to have spec tests for this.

**Comment by ajayaa** [comment]
please test this

File: manifests/keystone/role.pp [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+) {
+
+  keystone_role { $role:
```
Comment (bodepd) [inline manifests/keystone/role.pp]: The title here should be $name, not $role.

File: manifests/keystone/role.pp (hunk 1 above)
Comment (bodepd) [inline manifests/keystone/role.pp]: it seems like this should be omitted and you should just use $name as the identifier.

**Comment by bodepd** [comment]
last build failed b/c role was specified and not set, removing it and relying on $name will resolve the issue

````
Jul 15 19:01:55 ocdb1-testjenkins-puppet-rjil-gate-1783 puppet-user[14259]: Must pass role to Rjil::Keystone::Role[tenant_admin] at /etc/puppet/modules.overrides/rjil/manifests/openstack_objects.pp:82 on node ocdb1-testjenkins-puppet-rjil-gate-1783
````

**Comment by ajayaa** [comment]
please test this

**Comment by ajayaa** [comment]
please test this

**Comment by ajayaa** [comment]
please test this

**Comment by ajayaa** [comment]
please test this

**Comment by ajayaa** [comment]
please test this

**Comment by ajayaa** [comment]
please test this

File: manifests/keystone/role.pp [hunk 3]
```diff
@@ -0,0 +1,14 @@
+##
+# Define: rjil::keystone::role
+#
+# == Purpose: create Keystone roles
+#
+##
+define rjil::keystone::role (
```
Comment (hkumarmk) [inline manifests/keystone/role.pp]: why you need this type? you can directly call keystone_role from where this is calling

File: manifests/openstack_objects.pp [hunk 4]
```diff
@@ -77,4 +78,9 @@
     create_resources('rjil::keystone::tenants',$tenants)
   }
 
+  if is_array($roles) {
+    rjil::keystone::role { $roles: }
```
Comment (hkumarmk) [inline manifests/openstack_objects.pp]: you can just call keystone_role.

**Comment by ajayaa** [comment]
please test this

**Comment by ajayaa** [comment]
please test this

**Comment by ajayaa** [comment]
please test this

**Comment by ajayaa** [comment]
please test this

**Comment by ajayaa** [comment]
please test this

**Comment by ajayaa** [comment]
please test this

**Comment by ajayaa** [comment]
please test this

**Comment by ajayaa** [comment]
please test this

**Comment by ajayaa** [comment]
please test this

**Comment by ajayaa** [comment]
This patch builds properly when I bring the build up in my own tenant. That makes the puppet code syntax error free. Most of the times the initial puppet run itself fails. But since I can't log into the nodes if the initial puppet run fails, I don't know why it's failing in jenkins tenant.

**Comment by ajayaa** [comment]
yay! The jenkins God finally blesses my patch.

**Comment by ajayaa** [comment]
please test this

**Comment by hkumarmk** [comment]
I checked the build, but I dont see such role

root@ocdb1-testjenkins-puppet-rjil-gate-1889:~# keystone role-list
+----------------------------------+---------------+
|                id                |      name     |
+----------------------------------+---------------+
| 55f12510e29148a189c0e2f611189ae0 |     Member    |
| 9ed746aeaeb64638b1ef1bd17832a04f | ResellerAdmin |
| 9fe2ff9ee4384b1894a90878d3e92bab |    _member_   |
| 01dfaf55eb8944e58950f049fdc33c35 |     admin     |
+----------------------------------+---------------+

**Comment by hkumarmk** [comment]
Also I have a query on role named "engineering" - should this be named after a functional capability based on what this role supposed to be capable of? say for example, tenant_admin in itself people know that it has admin capabilities for particular tenant.

**Comment by ajayaa** [comment]
@hkumarmk Have you seen this bug https://bugs.launchpad.net/jio/+bug/1470453? The engineering role is to be assigned to engineers in Jiocloud to let them debug issues with our cloud. If you can come up with a better name, I will take it.

**Comment by hkumarmk** [comment]
@ajayaa I understand the motive, I just suggested is about the name of the role sounded like a group name, anyway I dont know what all capabilities this role is going to have so I dont know if I have a name for it :)

**Comment by ajayaa** [comment]
please test this

**Comment by ajayaa** [comment]
please test this

**Comment by ajayaa** [comment]
please test this

**Comment by ajayaa** [comment]
please test this

**Comment by ajayaa** [comment]
please test this

**Comment by ajayaa** [comment]
please test this

**Comment by anshprat** [comment]
please test this

**Comment by ajayaa** [comment]
please test this

**Comment by ajayaa** [comment]
please test this

**Comment by anshprat** [comment]
please test this

**Comment by anshprat** [comment]
root@ocdb1-testjenkins-puppet-rjil-gate-1959:~# openstack role list
INFO: urllib3.connectionpool Starting new HTTPS connection (1): identity.jiocloud.com
INFO: urllib3.connectionpool Starting new HTTPS connection (1): identity.jiocloud.com
+----------------------------------+---------------+
| ID                               | Name          |
+----------------------------------+---------------+
| 7e8ccf160a0e4e819df530f0ed1a3af3 | admin         |
| 87570b5648b44b55a17f1745140ee153 | ResellerAdmin |
| 9fe2ff9ee4384b1894a90878d3e92bab | _member_      |
| bbbceca31d2141179a24bedb2200a115 | engineering   |
| c06680e2b3584bb8ac745de542b5ba72 | Member        |
| c8b9b6b3f3c94b77b55ac6f61442865b | tenant_admin  |
+----------------------------------+---------------+

**Comment by hkumarmk** [comment]
This is not going to affect any functionality of the system, so good to merge

**Pull request merged**